src/storage/admin/AdminTaskManager.cpp

```diff
@@ -0,0 +1,171 @@
+    }
+
+    pool_->stop();
+    pool_->wait();
+
+    bgThread_->stop();
+    bgThread_->wait();
+
+    LOG(INFO) << "exit AdminTaskManager::shutdown()";
+}
+
+void AdminTaskManager::pickTaskThread() {
+    std::chrono::milliseconds interval{20};    // 20ms
+    while (!shutdown_) {
+        LOG(INFO) << "waiting for incoming task";
+        folly::Optional<TaskHandle> optTaskHandle{folly::none};
+        while (!optTaskHandle && !shutdown_) {
+            optTaskHandle = taskHandleQueue_.try_take_for(interval);
+        }
+
+        if (shutdown_) {
+            LOG(INFO) << "detect AdminTaskManager::shutdown()";
+            return;
+        }
+
+        auto handle = *optTaskHandle;
+        LOG(INFO) << folly::stringPrintf("dequeue task(%d, %d)",
+                                         handle.first, handle.second);
+        auto it = tasks_.find(handle);
+        if (it == tasks_.end()) {
+            LOG(ERROR) << folly::stringPrintf(
+                        "trying to exec non-exist task(%d, %d)",
+                        handle.first, handle.second);
+            continue;
+        }
+
+        std::shared_ptr<TaskExecContext> ctx = tasks_[handle];
+        auto errOrSubTasks = ctx->task->genSubTasks();
+        if (!nebula::ok(errOrSubTasks)) {
+            LOG(ERROR) << "genSubTasks() failed=" << static_cast<int>(nebula::error(errOrSubTasks));
+            ctx->task->finish(nebula::error(errOrSubTasks));
+            tasks_.erase(handle);
+            return;
+        }
+
+        auto subTasks = nebula::value(errOrSubTasks);
+        for (auto& subtask : subTasks) {
+            ctx->subtasks.add(subtask);
+        }
+
+        auto subTaskConcurrency = std::min(ctx->task->getConcurrentReq(),
+                                           static_cast<size_t>(FLAGS_max_concurrent_subtasks));
+        subTaskConcurrency = std::min(subTaskConcurrency, subTasks.size());
+        ctx->notFinishedThread = subTaskConcurrency;
+
+        LOG(INFO) << folly::stringPrintf("run task(%d, %d) in %zu thread",
+                                         handle.first, handle.second, ctx->notFinishedThread);
+        for (size_t i = 0; i < subTaskConcurrency; ++i) {
+            pool_->addTask(&AdminTaskManager::pickSubTask, this, handle);
+        }
+    }  // end while (!shutdown_)
+    LOG(INFO) << "AdminTaskManager::pickTaskThread(~)";
+}
+
+void AdminTaskManager::pickSubTask(TaskHandle handle) {
+    std::chrono::milliseconds take_dura{10};
+    ResultCode successful{ResultCode::SUCCEEDED};
+    auto ctx = tasks_[handle];
+    while (auto subTask = ctx->subtasks.try_take_for(take_dura)) {
+        if (ctx->cancelled) {
+            break;
+        }
+        ResultCode rc = subTask->invoke();
+        if (rc != ResultCode::SUCCEEDED) {
+            ctx->rc.compare_exchange_strong(successful, rc);
+            break;
+        }
+    }
+
+    bool taskFinished{false};
+    if (ctx->taskFinished.compare_exchange_strong(taskFinished, true)) {
```

Comment:
I am confused about the code here. 
Based on my understanding,  for each subTask,   after it is finished,  the subTask::onFinished callback will be invoked, right?
What you need to do is to check current task whether all subtasks are finished, if true,  Task::onFinished will be invoked, otherwise, pick up another subtask.